terminus-ui/drawer/src/drawer.component.ts

```diff
@@ -0,0 +1,343 @@
+} from '@terminus/ngx-tools';
+import {
+  fromEvent,
+  merge,
+  Observable,
+  Subject,
+} from 'rxjs';
+import {
+  debounceTime,
+  distinctUntilChanged,
+  filter,
+  map,
+  startWith,
+  take,
+  tap,
+} from 'rxjs/operators';
+import { tsDrawerAnimations } from './drawer-animations';
+
+// Result of the toggle promise that indicates the state of the drawer.
+export type TsDrawerToggleResult = 'open' | 'close';
+
+// Type of drawer display mode
+export type TsDrawerModes = 'overlay' | 'push';
+
+// Type of drawer position
+export type TsDrawerPosition = 'start' | 'end';
+
+// Configures whether drawers should use auto sizing by default.
+export const TS_DRAWER_DEFAULT_AUTOSIZE =
+  new InjectionToken<boolean>('TS_DRAWER_DEFAULT_AUTOSIZE', {
+    providedIn: 'root',
+    factory: TS_DRAWER_DEFAULT_AUTOSIZE_FACTORY,
+  });
+
+export function TS_DRAWER_DEFAULT_AUTOSIZE_FACTORY(): boolean {
+  return false;
+}
+
+
+/**
+ * This component corresponds to a drawer that can be opened on the drawer container.
+ */
+@Component({
+  selector: 'ts-drawer',
+  templateUrl: './drawer.component.html',
+  styleUrls: ['./drawer.component.scss'],
+  animations: [tsDrawerAnimations.transformDrawer],
+  host: {
+    'class': 'ts-drawer',
+    // must prevent the browser from aligning text based on value
+    '[attr.align]': 'null',
+    '[attr.role]': 'role',
+    '[class.ts-drawer-end]': 'position === "end"',
+    '[class.ts-drawer-overlay]': 'mode === "overlay"',
+    '[class.ts-drawer-push]': 'mode === "push"',
+    'tabIndex': '-1',
+    '[@transform]': `{
+        value: animationState,
+        params: {
+            collapsedSize: collapsedSize,
+            expandedSize: expandedSize
+        }
+    }`,
+  },
+  changeDetection: ChangeDetectionStrategy.OnPush,
+  encapsulation: ViewEncapsulation.None,
+  exportAs: 'tsDrawer',
+})
+export class TsDrawerComponent implements AfterContentChecked, OnDestroy {
+
+  public animationState: 'open-instant' | 'open' | 'void' = 'void';
+
+  // Emits whenever the drawer has started animating.
+  public animationStarted = new Subject<AnimationEvent>();
+
+  // Emits whenever the drawer is done animating.
+  public animationEnd = new Subject<AnimationEvent>();
+
+  // Emits when the component is destroyed.
+  private readonly destroyed = new Subject<void>();
+
+  // Whether the drawer is initialized. Used for disabling the initial animation.
+  private enableAnimations = false;
+
+  /**
+   * An observable that emits when the drawer mode changes. This is used by the drawer container to
+   * to know when the mode changes so it can adapt the margins on the content.
+   */
+  public readonly modeChanged = new Subject();
+
+  public get width(): number {
+    return this.elementRef.nativeElement ? (this.elementRef.nativeElement.offsetWidth || 0) : 0;
+  }
+
+  // aria role label, default to nothing
+  @Input()
+  public role = '';
+
+  // Collapsed drawer width
+  @Input()
+  public set collapsedSize(value: string) {
```

Comment:
For `collapsedSize` and `expandedSize` we should abstract out the default to a constant above, then if the user passes in `null` or `undefined` we can fall back to the default. (Our utility `isUnset` would work well here.)